File: client/lib/services/chat_repository.dart
import 'dart:async';
import 'package:uuid/uuid.dart';
import '../crypto/crypto_engine.dart';
import '../network/api_client.dart';
import '../network/socket_client.dart';
import '../database_service.dart';
import '../models/message_state.dart';

import 'connection_manager.dart';
import 'pending_message_queue.dart';
import 'message_sync_service.dart';
import 'receipt_service.dart';
import 'typing_service.dart';
import 'presence_service.dart';
import 'profile_resolver.dart';

class ChatRepository {
  final CryptoEngine cryptoEngine;
  final ApiClient apiClient;
  final String myUserId;

  late final SocketClient socketClient;
  late final ConnectionManager connectionManager;
  late final PendingMessageQueue pendingQueue;
  late final MessageSyncService syncService;
  late final ReceiptService receiptService;
  late final TypingService typingService;
  late final PresenceService presenceService;
  late final ProfileResolver profileResolver;

  StreamSubscription? _readyToSendSub;

  ChatRepository({
    required this.cryptoEngine,
    required this.apiClient,
    required this.myUserId,
    required String socketUrl,
    required String authToken,
  }) {
    socketClient = SocketClient(
      socketUrl: socketUrl,
      authToken: authToken,
      myUserId: myUserId,
    );
    
    connectionManager = ConnectionManager(socketClient);
    pendingQueue = PendingMessageQueue(socketClient);
    profileResolver = ProfileResolver(apiClient: apiClient);
    syncService = MessageSyncService(socketClient: socketClient, cryptoEngine: cryptoEngine, profileResolver: profileResolver);
    receiptService = ReceiptService(socketClient: socketClient, pendingQueue: pendingQueue);
    typingService = TypingService(socketClient: socketClient);
    presenceService = PresenceService(socketClient: socketClient, apiClient: apiClient);

    _readyToSendSub = pendingQueue.readyToSend.listen(_processOutgoingMessage);
  }

  Future<void> initialize() async {
    // Generate/register keys
    final payload = await cryptoEngine.generateInitialKeys();
    await apiClient.registerKeys(payload);

    // Connect socket
    await connectionManager.connect();
  }

  // Enqueue a message to be sent reliably
  Future<String> enqueueMessage(String recipientIdentifier, String plaintext, {String? msgId}) async {
    final finalMsgId = msgId ?? const Uuid().v4();
    String actualRecipientId = recipientIdentifier;

    // Resolve username if needed (for MVP/v1 we assume API resolve)
    if (!_isUuid(recipientIdentifier)) {
      try {
        final bundle = await apiClient.fetchPreKeysByUsername(recipientIdentifier);
        actualRecipientId = bundle['user_id'];
      } catch (e) {
        throw Exception('Could not resolve username: $e');
      }
    }

    // Save as pending immediately
    final Map<String, dynamic> msg = {
      'msg_id': finalMsgId,
      'recipient_id': actualRecipientId,
      'sender_id': myUserId,
      'plaintext': plaintext,
    };
    
    await DatabaseService.saveMessage(
      recipientId: actualRecipientId,
      senderId: myUserId,
      text: plaintext,
      timestamp: DateTime.now(),
      msgId: finalMsgId,
      status: MessageState.pending,
      isMe: true,
    );

    await pendingQueue.enqueue(msg);
    return finalMsgId;
  }

  // Internal processor for the pending queue
  Future<void> _processOutgoingMessage(Map<String, dynamic> msg) async {
    final actualRecipientId = msg['recipient_id'];
    final plaintext = msg['plaintext'];
    final msgId = msg['msg_id'];

    try {
      Map<String, dynamic> encryptedPayload;
      try {
        encryptedPayload = await cryptoEngine.encryptMessage(actualRecipientId, plaintext);
      } catch (e) {
        // Fetch bundle and retry
        final bundle = await apiClient.fetchPreKeys(actualRecipientId);
        await cryptoEngine.establishSession(actualRecipientId, bundle);
        encryptedPayload = await cryptoEngine.encryptMessage(actualRecipientId, plaintext);
      }

      final response = await socketClient.sendMessage(actualRecipientId, msgId, {
        'sender_id': myUserId,
        ...encryptedPayload
      });

      // Server returns {status: 'delivered', server_ts: 123, seq: 456}
      await pendingQueue.onAck(msgId, response['server_ts'] as int, response['seq'] as int);

      // Update Conversation Summary
      final existing = DatabaseService.getContactProfile(actualRecipientId);
      await DatabaseService.updateConversation(
        userId: actualRecipientId,
        name: existing?['name'] ?? actualRecipientId,
        avatarUrl: existing?['avatarUrl'],
        lastMessage: plaintext,
        lastMessageTime: DateTime.fromMillisecondsSinceEpoch(response['server_ts'] as int),
      );

    } catch (e) {
      await pendingQueue.onTimeout(msgId);
    }
  }

  bool _isUuid(String input) {
    final RegExp regex = RegExp(r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$', caseSensitive: false);
    return regex.hasMatch(input);
  }

  void dispose() {
    _readyToSendSub?.cancel();
    pendingQueue.dispose();
    syncService.dispose();
    receiptService.dispose();
    typingService.dispose();
    presenceService.dispose();
    connectionManager.dispose();
  }
}

File: android_client/app/src/main/java/com/subhankar/aurachat/service/ChatRepository.kt
package com.subhankar.aurachat.service

import android.util.Log
import com.subhankar.aurachat.crypto.CryptoEngine
import com.subhankar.aurachat.data.local.dao.ContactDao
import com.subhankar.aurachat.data.local.dao.ConversationDao
import com.subhankar.aurachat.data.local.dao.MessageDao
import com.subhankar.aurachat.data.local.entity.ConversationEntity
import com.subhankar.aurachat.data.local.entity.MessageEntity
import com.subhankar.aurachat.data.local.entity.MessageStatus
import com.subhankar.aurachat.data.local.SessionManager
import com.subhankar.aurachat.network.ApiClient
import com.subhankar.aurachat.network.AppConfig
import com.subhankar.aurachat.network.ConnectionManager
import com.subhankar.aurachat.network.PhoenixSocketManager
import java.time.Instant
import java.util.UUID
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Chat Repository — replaces Dart chat_repository.dart.
 *
 * This is the single orchestrator that wires together:
 *   • CryptoEngine (encryption/decryption)
 *   • PhoenixSocketManager (WebSocket)
 *   • ConnectionManager (reconnection)
 *   • PendingMessageQueue (reliable delivery)
 *   • MessageSyncService (incoming messages)
 *   • ReceiptService (delivery/read receipts)
 *   • TypingService (typing indicators)
 *   • PresenceService (online/offline status)
 *
 * In Dart, the ServiceLocator created all these manually.
 * In Kotlin, Hilt handles dependency injection automatically.
 */
@Singleton
class ChatRepository @Inject constructor(
    private val cryptoEngine: CryptoEngine,
    private val apiClient: ApiClient,
    private val socketManager: PhoenixSocketManager,
    private val connectionManager: ConnectionManager,
    private val pendingQueue: PendingMessageQueue,
    private val syncService: MessageSyncService,
    private val receiptService: ReceiptService,
    private val typingService: TypingService,
    private val presenceService: PresenceService,
    private val sessionManager: SessionManager,
    private val messageDao: MessageDao,
    private val conversationDao: ConversationDao,
    private val contactDao: ContactDao
) {
    private val repositoryScope = kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO)

    companion object {
        private const val TAG = "ChatRepository"
    }

    /**
     * Initialize the entire messaging system.
     * Replaces: Dart chatRepository.initialize()
     *
     * Call this once after login/auth is complete.
     */
    suspend fun initialize() {
        // 1. Generate and register E2EE keys with the server
        val keyBundle = cryptoEngine.generateInitialKeys()
        apiClient.registerKeys(keyBundle)
        Log.d(TAG, "E2EE keys registered")

        // 2. Connect WebSocket
        val userId = sessionManager.getUserId() ?: throw IllegalStateException("Not logged in")
        val token = sessionManager.getToken() ?: throw IllegalStateException("No auth token")
        connectionManager.connect(
            socketUrl = AppConfig.socketEndpoint,
            token = token,
            userId = userId
        )

        // 3. Start all services
        syncService.start()
        receiptService.start()
        typingService.start()
        presenceService.start()

        // 4. Retry any messages that were pending from a previous session
        pendingQueue.retryAll()

        Log.d(TAG, "ChatRepository initialized — all services running")
    }

    /**
     * Send a message reliably.
     * Replaces: Dart chatRepository.enqueueMessage()
     *
     * Flow:
     *   1. Save to local DB immediately (shows in UI with "pending" clock icon)
     *   2. Enqueue for reliable delivery (encryption + send + retry)
     *   3. On ACK: update to "sent" (single tick)
     *   4. On delivery receipt: update to "delivered" (double tick)
     *   5. On read receipt: update to "read" (blue double tick)
     */
    suspend fun sendMessage(recipientId: String, plaintext: String, msgId: String? = null): String {
        val finalMsgId = msgId ?: UUID.randomUUID().toString()
        val myUserId = sessionManager.getUserId() ?: throw IllegalStateException("Not logged in")

        if (recipientId == "dummy_user_123") {
            // Save user message to local DB immediately as READ
            val message = MessageEntity(
                msgId = finalMsgId,
                chatId = recipientId,
                senderId = myUserId,
                text = plaintext,
                timestamp = Instant.now().toString(),
                isMe = true,
                status = MessageStatus.READ
            )
            messageDao.insert(message)

            // Update conversation summary
            conversationDao.upsert(
                ConversationEntity(
                    userId = recipientId,
                    name = "AuraChat Assistant",
                    avatarUrl = null,
                    lastMessage = plaintext,
                    lastMessageTime = Instant.now().toString(),
                    unreadCount = 0
                )
            )

            // Simulate reply from the assistant in the background
            /*
            repositoryScope.launch {
                try {
                    kotlinx.coroutines.delay(1000)
                    val responseText = getAssistantResponse(plaintext)
                    val responseMsgId = UUID.randomUUID().toString()
                    val responseMessage = MessageEntity(
                        msgId = responseMsgId,
                        chatId = recipientId,
                        senderId = recipientId,
                        text = responseText,
                        timestamp = Instant.now().toString(),
                        isMe = false,
                        status = MessageStatus.READ
                    )
                    messageDao.insert(responseMessage)

                    conversationDao.upsert(
                        ConversationEntity(
                            userId = recipientId,
                            name = "AuraChat Assistant",
                            avatarUrl = null,
                            lastMessage = responseText,
                            lastMessageTime = Instant.now().toString(),
                            unreadCount = 0
                        )
                    )
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            */
            return finalMsgId
        }

        // Save to local DB immediately — appears in chat with pending status
        val message = MessageEntity(
            msgId = finalMsgId,
            chatId = recipientId,
            senderId = myUserId,
            text = plaintext,
            timestamp = Instant.now().toString(),
            isMe = true,
            status = MessageStatus.PENDING
        )
        messageDao.insert(message)

        // Update conversation summary
        val contact = contactDao.getContact(recipientId)
        conversationDao.upsert(
            ConversationEntity(
                userId = recipientId,
                name = contact?.name ?: recipientId,
                avatarUrl = contact?.avatarUrl,
                lastMessage = plaintext,
                lastMessageTime = Instant.now().toString()
            )
        )

        // Enqueue for reliable delivery
        pendingQueue.enqueue(
            msgId = finalMsgId,
            recipientId = recipientId,
            senderId = myUserId,
            plaintext = plaintext
        )

        return finalMsgId
    }

    private fun getAssistantResponse(input: String): String {
        val lower = input.lowercase().trim()
        return when {
            lower.contains("hello") || lower.contains("hi") || lower.contains("hey") -> {
                "Hello there! I am your AuraChat Assistant. How can I help you customize your chat screen today?"
            }
            lower.contains("color") || lower.contains("theme") || lower.contains("custom") -> {
                "AuraChat has premium colors built-in! Tell me which colors or fonts you want to customize on the Chat screen."
            }
            lower.contains("help") -> {
                "I can assist with testing messages, checking bubble layouts, and validating styling. Just type anything here!"
            }
            else -> {
                "Awesome! You typed: \"$input\". Tell me how you want to design or customize this chat bubble interface."
            }
        }
    }

    /**
     * Called when app returns to foreground.
     */
    fun onAppResumed() {
        connectionManager.onAppResumed()
    }

    /**
     * Clean shutdown.
     */
    fun dispose() {
        syncService.dispose()
        receiptService.dispose()
        typingService.dispose()
        presenceService.dispose()
        pendingQueue.dispose()
        connectionManager.dispose()
    }
}
